### PR TITLE
Make custom trait object for `Future` generic

### PR DESCRIPTION
**Pull request by MajorBreakfast**
- `TaskObj` -> `FutureObj<'static, ()>`
- The `impl From<...> for FutureObj<'a, T>` impls are impossible because of the type parameter `T`. The impl has to live in libstd, but `FutureObj<'a, T>` is from libcore. Therefore `Into<FutureObj<'a, T>>` was implemented instead. Edit: This didn‘t compile without warnings. I am now using non-generic Form impls.

See https://github.com/rust-lang-nursery/futures-rs/issues/1058

r? @cramertj 

Edit: Added lifetime

### CODE REVIEW

**Comment by ngg** [comment]
I get the following warnings for the `Into` implementations:
```
warning: conflicting implementations of trait `core::convert::Into<core::task::FutureObj<_>>` for type `boxed::PinBox<_>`: (E0119)
   --> liballoc/boxed.rs:953:1
    |
953 | impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for PinBox<F> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: #[warn(incoherent_fundamental_impls)] on by default
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #46205 <https://github.com/rust-lang/rust/issues/46205>
    = note: conflicting implementation in crate `core`:
            - impl<T, U> core::convert::Into<U> for T
              where U: core::convert::From<T>;
    = note: downstream crates may implement trait `core::convert::From<boxed::PinBox<_>>` for type `core::task::FutureObj<_>`
```
Can't we simply remove that impl?

**Comment by MajorBreakfast** [comment]
@nikomatsakis 
@aturon 
@withoutboats 

Can you help? I'm not sure what to do. I can implement neither  `From` nor `Into`

**Comment by MajorBreakfast** [comment]
I've removed the `Into` impls again and reverted to `From` impls, but just for `FutureObj<()>` instead of `FutureObj<T>`

<details>
<summary>Into impls that I removed</summary>

```rust
#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for PinBox<F> {
    fn into(self) -> FutureObj<T> {
        FutureObj::new(self)
    }
}

#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + Send + 'static> Into<FutureObj<T>> for Box<F> {
    fn into(self) -> FutureObj<T> {
        FutureObj::new(PinBox::from(self))
    }
}

#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for PinBox<F> {
    fn into(self) -> LocalFutureObj<T> {
        LocalFutureObj::new(self)
    }
}

#[unstable(feature = "futures_api", issue = "50547")]
impl<T, F: Future<Output = T> + 'static> Into<LocalFutureObj<T>> for Box<F> {
    fn into(self) -> LocalFutureObj<T> {
        LocalFutureObj::new(PinBox::from(self))
    }
}
```
</details>

**Comment by MajorBreakfast** [comment]
- Added lifetime
- Moved `FuturesObj` to `core::future`

**Comment by ngg** [comment]
I just tried to use it to create async trait methods (including capturing &self with the generic lifetime), it seems to work great.

**Comment by MajorBreakfast** [comment]
I've added an impl for `PinMut`. It's possible now that `FutureObj` can store a lifetime.

@ngg Thats' great to hear!

**Comment by ngg** [comment]
Actually it seems to work properly, it may be sound.

**Comment by ngg** [comment]
For example `futures::executor::block_on(obj)` doesn't require `obj` to be `'static`.

**Comment by MajorBreakfast** [comment]
@ngg You're right. `FutureObj` with non-static lifetime is definitely useful with `block_on`. I've just tested this and it works with the non-static `FutureObj` I created.

@cramertj 
- Do you think that a `FutureObj` impl for `PinMut` makes sense conceptually?
- I think that `UnsafeFutureObj` should be implemented for `Box` as well. A `PinBox` is IMO not necessary because the future isn't pinned at that point.

**Comment by MajorBreakfast** [comment]
@ngg Thanks BTW for all your testing :)

File: src/libcore/future/future_obj.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+use task::{Context, Poll};
+
+/// A custom trait object for polling futures, roughly akin to
+/// `Box<dyn Future<Output = T> + 'a>`.
+/// Contrary to `FutureObj`, `LocalFutureObj` does not have a `Send` bound.
+pub struct LocalFutureObj<'a, T> {
+    ptr: *mut (),
+    poll_fn: unsafe fn(*mut (), &mut Context) -> Poll<T>,
+    drop_fn: unsafe fn(*mut ()),
+    _marker1: PhantomData<T>,
```
Comment (cramertj) [inline src/libcore/future/future_obj.rs]: What is this marker for? `T` is already used in the signature of `poll_fn`.

**Comment by cramertj** [comment]
@MajorBreakfast Thanks as always for this PR!
> Do you think that a FutureObj impl for PinMut makes sense conceptually?

Yeah, this is great! It's analogous to `PinMut<'a, dyn Future>` or `PinMut<'a, dyn Future + Send>`. I would also add an impl for `&mut T` where `T: Future + Unpin`.

**Comment by cramertj** [comment]
r=me with `PhantomType` nit addressed and `&mut` impl added.

**Comment by mikeyhew** [comment]
@cramertj they will probably be in nightly within the next month, but not sure when they will be stabilized

**Comment by cramertj** [comment]
@mikeyhew Cool! Thanks again for working on that.

@MajorBreakfast I think with that timeline it's still best for us to push forward with this approach and then we can adjust as necessary-- we need both unsized rvalues and object-safe arbitrary self types before we can totally replace these faked-out trait objects.

File: src/libcore/future/future_obj.rs (hunk 1 above)
Comment (MajorBreakfast) [reply src/libcore/future/future_obj.rs]: It's for uhm... *kicks it under the rug* ...what marker do you mean? :)

**Comment by cramertj** [comment]
Great!

@bors r+

**Comment by ngg** [comment]
@cramertj Are you sure that `PhantomData<T>` is really not necessary? This struct does own `T` in the sense that its `Drop` implementation will drop `T`, so I think it's necessary for dropck. (`poll_fn` should be enough for borrowck)

**Comment by MajorBreakfast** [comment]
@ngg AFAIK `PhantomData`'s only purpose is to avoid `error[E0392]: parameter T is never used`

**Comment by MajorBreakfast** [comment]
@ngg I don't think it is accurate to say that the "struct does own `T`" because `T` is only a type and therefore can't be owned because it's not something that is stored in memory.

**Comment by ngg** [comment]
It will not drop `T` when dropped, so my comment was kinda meaningless, sry.
@MajorBreakfast `PhantomData` does lots of other things, it's not just for avoding an error, see https://doc.rust-lang.org/nomicon/phantom-data.html and specifically for dropck uses see https://stackoverflow.com/questions/42708462/why-is-it-useful-to-use-phantomdata-to-inform-the-compiler-that-a-struct-owns-a/42721125#42721125

**Comment by cramertj** [comment]
@ngg `PhantomData` isn't needed here-- the type `T` is already used as the output of the `poll_fn`, so that supplies the variance required. The struct itself implements `Drop` manually (always) regardless of the inner type `T`, so adding the `T` here doesn't affect drop checking (AFAIK).

**Comment by MajorBreakfast** [comment]
@cramertj I'm not so sure... I think @ngg is right an we're violating this

Here's what `Vec` uses internally:
https://doc.rust-lang.org/1.26.0/src/core/ptr.rs.html#2514-2522

**Comment by ngg** [comment]
Yeah that's what I meant but it wouldn't matter to add `PhantomData<T>` because it's not `T` what we own, but the original Future (let's call it `F`). This whole FutureObj is just emulating `Box<Future<Output = T> + 'a>` which would contain `PhantomData<Future<Output = T> + 'a>` so probably that is what we need here. (I'm really not sure about these things, I'm just guessing.. It might be ok now)

**Comment by cramertj** [comment]
@ngg Yeah, if you could spell `PhantomData<Box<dyn UnsafeFutureObj<'a, T>>`, that would be correct, but we don't own a `T`, so `PhantomData<T>` doesn't do anything for us.

**Pull request merged**